Lib/inspect.py

```diff
@@ -2283,7 +2283,9 @@ def _signature_fromstr(cls, obj, s, skip_bound_arg=True):
 
     module = None
     module_dict = {}
-    module_name = getattr(obj, '__module__', None)
+    module_name = (getattr(obj, '__module__', None) or
+                   getattr(getattr(obj, '__objclass__', None),
+                           '__module__', None))
```

Comment:
Since the expression is too long and complex so we need to wrap the lines, I would split it into several simpler statements, for example:
```py
    module_name = getattr(obj, '__module__', None)
    if not module_name:
        objclass = getattr(obj, '__objclass__', None)
        module_name = getattr(objclass, '__module__', None)
```
or
```py
    module_name = getattr(obj, '__module__', None)
    if not module_name:
        objclass = getattr(obj, '__objclass__', None)
        if objclass is not None:
            module_name = objclass.__module__
```